File: python/interview/classic/thread_pool/thread_pool.py
"""简易线程池：`queue` + 固定数量 worker 线程；`shutdown` 先 `join` 再发哨兵。"""

from __future__ import annotations

import threading
from queue import Queue
from typing import Any, Callable, Optional, Tuple


class ThreadPool:
    def __init__(self, num_workers: int) -> None:
        if num_workers < 1:
            raise ValueError("num_workers must be >= 1")
        self._q: Queue[Optional[Tuple[Callable[..., Any], tuple[Any, ...], dict[str, Any]]]] = Queue()
        self._workers: list[threading.Thread] = []
        self._started = True
        for _ in range(num_workers):
            t = threading.Thread(target=self._loop, daemon=True)
            self._workers.append(t)
            t.start()

    def _loop(self) -> None:
        while True:
            item = self._q.get()
            try:
                if item is None:
                    return
                fn, args, kwargs = item
                fn(*args, **kwargs)
            finally:
                self._q.task_done()

    def submit(self, fn: Callable[..., Any], *args: Any, **kwargs: Any) -> None:
        if not self._started:
            raise RuntimeError("pool is shut down")
        self._q.put((fn, args, kwargs))

    def shutdown(self, wait: bool = True) -> None:
        if not self._started:
            return
        self._started = False
        if wait:
            self._q.join()
        for _ in self._workers:
            self._q.put(None)
        for t in self._workers:
            t.join()


if __name__ == "__main__":
    lock = threading.Lock()
    acc: list[int] = []

    def work(x: int) -> None:
        with lock:
            acc.append(x)

    pool = ThreadPool(4)
    for i in range(40):
        pool.submit(work, i)
    pool.shutdown(wait=True)
    assert len(acc) == 40
    assert set(acc) == set(range(40))
    print("thread_pool OK")

File: cpp/interview/classic/thread_pool/thread_pool.cpp
// 简易线程池：mutex + condition_variable + queue<function<void()>>
#include <bits/stdc++.h>
#include <cassert>
#include <condition_variable>
#include <functional>
#include <mutex>
#include <queue>
#include <thread>
using namespace std;

class ThreadPool {
public:
    explicit ThreadPool(size_t n) : stop_(false), joined_(false) {
        for (size_t i = 0; i < n; ++i) {
            workers_.emplace_back([this] { loop(); });
        }
    }

    ~ThreadPool() { shutdown(); }

    void enqueue(function<void()> fn) {
        {
            lock_guard<mutex> lk(mtx_);
            if (stop_) throw runtime_error("pool stopped");
            tasks_.push(move(fn));
        }
        cv_.notify_one();
    }

    void wait_idle() {
        unique_lock<mutex> lk(mtx_);
        idle_cv_.wait(lk, [&] { return pending_ == 0 && tasks_.empty(); });
    }

    void shutdown() {
        unique_lock<mutex> guard(join_mtx_);
        if (joined_) return;
        {
            unique_lock<mutex> lk(mtx_);
            idle_cv_.wait(lk, [&] { return pending_ == 0 && tasks_.empty(); });
            stop_ = true;
        }
        cv_.notify_all();
        for (auto& t : workers_) {
            if (t.joinable()) t.join();
        }
        joined_ = true;
    }

private:
    vector<thread> workers_;
    queue<function<void()>> tasks_;
    mutex mtx_;
    mutex join_mtx_;
    condition_variable cv_;
    condition_variable idle_cv_;
    int pending_ = 0;
    bool stop_;
    bool joined_;

    void loop() {
        while (true) {
            function<void()> task;
            {
                unique_lock<mutex> lk(mtx_);
                cv_.wait(lk, [&] { return stop_ || !tasks_.empty(); });
                if (stop_ && tasks_.empty()) return;
                task = move(tasks_.front());
                tasks_.pop();
                ++pending_;
            }
            task();
            {
                lock_guard<mutex> lk(mtx_);
                --pending_;
                if (pending_ == 0 && tasks_.empty()) idle_cv_.notify_all();
            }
        }
    }
};

int main() {
    mutex m;
    vector<int> acc;
    {
        ThreadPool pool(4);
        for (int i = 0; i < 40; ++i) {
            pool.enqueue([i, &m, &acc] {
                lock_guard<mutex> lk(m);
                acc.push_back(i);
            });
        }
        pool.wait_idle();
    }
    assert((int)acc.size() == 40);
    sort(acc.begin(), acc.end());
    for (int i = 0; i < 40; ++i) assert(acc[i] == i);
    cout << "thread_pool OK" << endl;
    return 0;
}
